### PR TITLE
fix(DB/creature_formations): Moon Priestess Amara pathing

### LINKED ISSUE #2700: [NPC] Moon Priestess Amara
##### SMALL DESCRIPTION:

NPCs are not linked 
InkedWoWScrnShot_022320_180012_LI

##### EXPECTED BLIZZLIKE BEHAVIOUR:

They have to walk together, the one in the white frame always in the middle. If we talk to one of them, they all stop. As in this video, unfortunately, I couldn't find any video from the WOTLK era but nothing has changed since then.
https://youtu.be/gAFeMHS9b-8?t=135

##### CURRENT BEHAVIOUR:

NPCs are not linked 
https://wowgaming.altervista.org/aowow/?npc=2151

##### STEPS TO REPRODUCE THE PROBLEM:

1.  .go cr 2151
2. 
3. 

##### BRANCH(ES):

master

### PR DESCRIPTION
**Pull request by LukasVolgger**
Use creature_formations instead of pathing for each npc. This avoids that they get out of sync after a while or get stuck.

I didn't change the formation, just rebuild the positions in formation as it was before - see screenshots below! 

**Before:** 
--------
image

**After:**
--------
image

## CHANGES PROPOSED:
-  update paths for creature 46394 & 46416 so only the path of 46414 is used
-  create formation for guid 46414 

## ISSUES ADDRESSED:
- Closes #2700 

## TESTS PERFORMED:

- Imported SQL with HeidiSQL on Win 10 Pro x64
- Tested successfully ingame

## HOW TO TEST THE CHANGES:

- .go c 46414
- Follow the formation a while and check that they are synced
- If you talk to any npc the whole formation should stop

## KNOWN ISSUES AND TODO LIST:

- [ ]
- [ ] 

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by Endeffect-AT** [comment]
Just kept the waypoint data as backup. Now gone :)

**Comment by Stoabrogga** [comment]
Thanks. We don't need a backup, try to keep the DB clean. If really necessary we could always restore the paths from an older git commit.

**Comment by Stoabrogga** [comment]
I pushed a commit to update the MovementType of the creatures, otherwise you get a DB error (see Travis log).

**Comment by Maurowin** [comment]
To me, doesn't work
WoWScrnShot_022520_231510

**Comment by Maurowin** [comment]
On the whole it works, but when I talk to the white one the others keep walking. Then a few seconds later, she leaves to catch up with the others. Normally, she would have to stop everything.

WoWScrnShot_022520_231903

**Comment by Stoabrogga** [comment]
I also tested this, and although the formation is moving around, they don't stop if talking to Moon Priestess Amara. The reason for this is that Huntress Nhemai (GUID 46414) is leading the formation and not Moon Priestess Amara (GUID 46394). @Endeffect-AT Could you please change this? And please also test this using an Alliance character talking to Amara.

**Comment by Maurowin** [comment]
> 
> 
> I also tested this, and although the formation is moving around, they don't stop if talking to Moon Priestess Amara. The reason for this is that Huntress Nhemai (GUID 46414) is leading the formation and not Moon Priestess Amara (GUID 46394). @Endeffect-AT Could you please change this? And please also test this using an Alliance character talking to Amara.

i had uploaded video on YT

**Comment by Endeffect-AT** [comment]
I will update this today. First i had the priestess as leader but every time the formation breaks. Also guid 46414 have a better path than 46394 - but it can be changed so that 46394 can use this.

**Comment by Stoabrogga** [comment]
You can easily take over the other waypoints by using this in your SQL script:
```sql
-- Delete waypoints for creatures 46394 and 46416
DELETE FROM `waypoint_data` WHERE `id` IN (463940,464160);

-- Take over waypoints from creature 46414 to creature 46394
INSERT INTO `waypoint_data` (`id`, `point`, `position_x`, `position_y`, `position_z`, `orientation`, `delay`, `move_type`, `action`, `action_chance`, `wpguid`)
SELECT 463940, `point`, `position_x`, `position_y`, `position_z`, `orientation`, `delay`, `move_type`, `action`, `action_chance`, `wpguid` FROM `waypoint_data` WHERE `id` = 464140;

-- Delete waypoints for creature 46414
DELETE FROM `waypoint_data` WHERE `id` = 464140;
```

**Comment by Volgger** [comment]
So. Now i tried it with leader 46394 but the npc also stops if a player interacts with her and ruins the whole formation for a few minutes. 

How to avoid that? I have no idea at all...
What you say about that look of the formation 

WoWScrnShot_022620_202539
WoWScrnShot_022620_202546
WoWScrnShot_022620_202602

**Comment by Stoabrogga** [comment]
I updated the SQL script and added the waypoints directly (I think it was a bad idea to take them over, because this only works once, you cannot execute the script multiple times, sorry for that). Also the waypoints for Amara have to be deleted before the insert. I tested the script and it works quite good in my opinion. After talking to Amara her companions stop a little bit later and after 2 minutes she continues with her patrol.

**Comment by Maurowin** [comment]
> 
> 
> I updated the SQL script and added the waypoints directly (I think it was a bad idea to take them over, because this only works once, you cannot execute the script multiple times, sorry for that). Also the waypoints for Amara have to be deleted before the insert. I tested the script and it works quite good in my opinion. After talking to Amara her companions stop a little bit later and after 2 minutes she continues with her patrol catching up with her formation members.

Do u think it's fixed in relation to what is said Volgger ? I will test that tonight.

**Comment by Stoabrogga** [comment]
Amara's companions still run a bit ahead after talking with Amara (depends on where the next waypoint is) but they will wait until Amara catches up again. So in my opinion this is much better than it was before where the companions just continued to follow their path on their own.

**Comment by Volgger** [comment]
I restored the version tag and have adjusted the positions. I agree that it's much better now than before. 

Thx @Stoabrogga for your help, nice PR to learn about pathing! 👍

**Comment by Stoabrogga** [comment]
> I restored the version tag

Great, thanks. I accidentally removed this, sorry.

**Comment by Stoabrogga** [comment]
@Volgger 
There are different kinds of waypoint paths. If you are interested there's a description on our wiki I wrote some time ago: http://www.azerothcore.org/wiki/Waypoints-Information

**Comment by Volgger** [comment]
Thx! I already studied your wiki, it's very helpful - Good work!

**Comment by Maurowin** [comment]
Can i test?

**Comment by Volgger** [comment]
Yes @Maurowin

**Comment by FrancescoBorzi** [comment]
@Maurowin you can always test everything that has the "Waiting to be tested" label 

image

**Comment by pklloveyou** [comment]
```sql
UPDATE `creature` SET `MovementType` = 0 WHERE `guid` IN 
(SELECT `memberGUID` FROM creature_formations WHERE `leaderGUID` != `memberGUID`);

UPDATE `creature_addon` SET `path_id` = 0 WHERE `guid` IN 
(SELECT `memberGUID` FROM creature_formations WHERE `leaderGUID` != `memberGUID`);

DELETE FROM `waypoint_data` WHERE `id` IN 
((SELECT `memberGUID` FROM creature_formations WHERE `leaderGUID` != `memberGUID`) + 0);
```

**Comment by Maurowin** [comment]
@FrancescoBorzi Yes I know, but since there were changes during the conversation, I wanted to know if it was updated before testing. I don't want to test the same code twice.

**Comment by pklloveyou** [comment]
How should I write "+0" in the third statement

**Comment by pklloveyou** [comment]
Don't do that. I'm just here to ask questions

**Comment by Volgger** [comment]
Thx for that! Good script tu use negation ainstead of values! But why setting `spawndist` = 0? if the `MovementType` is set to 0 they shouldn move at all from spawnpoint?

**Comment by pklloveyou** [comment]
Spawndist not 0 server console will report an error

**Comment by Volgger** [comment]
Ok, i check this out later!
@Maurowin you can test this for now as it is, i will check the modification from @pklloveyou tonight - but it would change nothing in the formation.

**Comment by pklloveyou** [comment]
I just want to clean up the database,

File: data/sql/updates/pending_db_world/rev_1582663580844407800.sql [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+(463940,18,9907.7,1206.18,1308.46,0,0,0,0,100,0),
+(463940,19,9916.67,1166.02,1307.9,0,0,0,0,100,0),
+(463940,20,9907.02,1133.67,1307.95,0,0,0,0,100,0),
+(463940,21,9881.11,1108.69,1307.99,0,0,0,0,100,0),
+(463940,22,9868.7,1077.61,1307.99,0,0,0,0,100,0),
+(463940,23,9848.29,1050.17,1305.27,0,0,0,0,100,0),
+(463940,24,9845.39,1007.59,1305.43,0,0,0,0,100,0);
+
+-- Delete waypoints for creature 46414 and 46416
+DELETE FROM `waypoint_data` WHERE `id` IN (464140, 464160);
```
Comment (Volgger) [inline data/sql/updates/pending_db_world/rev_1582663580844407800.sql]: They are already deleted here

**Comment by pklloveyou** [comment]
Remember to back them up before testing

**Comment by pklloveyou** [comment]
Its changes are changes to all members

**Comment by Stoabrogga** [comment]
@pklloveyou 
Spawndist is already 0 for those mobs, because they originally used 2 as movement type (waypoint movement). Spawndist only has to be set if movement type 1 (random movement) is used.

The SQL script is good as it is now, no error in the server log.

**Comment by pklloveyou** [comment]
Yes, but I'm talking about the statement I mentioned above

**Comment by Volgger** [comment]
I backup every single table which i modify in my PR's, also all tables are backed up.

**Comment by Stoabrogga** [comment]
@pklloveyou 
We don't need the waypoint data anymore for the two formation members, so they should be deleted.

**Comment by pklloveyou** [comment]
```sql
DELETE FROM `waypoint_data` WHERE `id` IN 
((SELECT `memberGUID` FROM creature_formations WHERE `leaderGUID` != `memberGUID`) * 10);
```
That doesn't sound right. Execution times are wrong

**Comment by Stoabrogga** [comment]
> 
> 
> ```sql
> DELETE FROM `waypoint_data` WHERE `id` IN 
> ((SELECT `memberGUID` FROM creature_formations WHERE `leaderGUID` != `memberGUID`) * 10);
> ```
> 
> That doesn't sound right. Execution times are wrong

Where did you get this SQL statement? It's not part of this PR.

**Comment by pklloveyou** [comment]
I want to clean up waypoint_data for all fleet members

**Comment by Stoabrogga** [comment]
@pklloveyou 
Sorry, but this discussion is about the PR, not for general support for cleaning up the DB.

**Comment by pklloveyou** [comment]
Ok, thank you, I'll delete the above statement,

**Comment by Stoabrogga** [comment]
@pklloveyou 
Please don't delete your comments. I'll hide them as "off-topic", as otherwise nobody knows what's going on here.

**Comment by pklloveyou** [comment]
okay

**Comment by pklloveyou** [comment]
```sql
UPDATE `creature` SET `MovementType` = 0, `spawndist` = 0 WHERE `guid` IN 
(SELECT `memberGUID` FROM creature_formations WHERE `leaderGUID` != `memberGUID`);

UPDATE `creature_addon` SET `path_id` = 0 WHERE `guid` IN 
(SELECT `memberGUID` FROM creature_formations WHERE `leaderGUID` != `memberGUID`);
```
```sql
DELETE FROM `waypoint_data` WHERE `id` IN 
((SELECT `memberGUID` FROM creature_formations WHERE `leaderGUID` != `memberGUID`) * 10);
```
This statement is to clean up all the members of the formation waypoint_data，But it has errors,

**Comment by Stoabrogga** [comment]
I tested this PR and it works very well. Can be merged in my opinion. Did follow Amara around for the full path, talkinig to her at different times. She always did catch up with her companions.

**Pull request merged**